src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -458,69 +496,229 @@ pub trait InterruptHandle: Debug + Send + Sync {
-    /// This is used to
-    /// 1. make sure stale signals do not interrupt the
-    ///    the wrong vcpu (a vcpu may only be interrupted iff `cancel_requested` is true),
-    /// 2. ensure that if a vm is killed while a host call is running,
-    ///    the vm will not re-enter the guest after the host call returns.
-    cancel_requested: AtomicBool,
-    /// True when the debugger has requested the VM to be interrupted. Set immediately when
-    /// `kill_from_debugger()` is called, and cleared when the vcpu is no longer running.
-    /// This is used to make sure stale signals do not interrupt the the wrong vcpu
-    /// (a vcpu may only be interrupted by a debugger if `debug_interrupt` is true),
+
+    /// Generation-aware cancellation request flag.
+    ///
+    /// **Bit 63**: Cancellation requested flag (1 = kill requested, 0 = no kill)
+    /// **Bits 0-62**: Generation number when cancellation was requested
+    ///
+    /// # Purpose
+    ///
+    /// This flag serves three critical functions:
+    ///
+    /// 1. **Prevent stale signals**: A VCPU may only be interrupted if cancel_requested
+    ///    is set AND the generation matches the current call's generation
+    ///
+    /// 2. **Handle host function calls**: If kill() is called while a host function is
+    ///    executing (VCPU not running but call is active), cancel_requested is stamped
+    ///    with the current generation. When the host function returns and the VCPU
+    ///    attempts to re-enter the guest, it will see the cancellation and abort.
+    ///
+    /// 3. **Detect stale kills**: If cancel_requested.generation doesn't match the
+    ///    current generation, it's from a previous call and should be ignored
+    ///
+    /// # States and Transitions
+    ///
+    /// - **No cancellation**: cancel_requested = 0 (bit 63 clear)
+    /// - **Cancellation for generation N**: cancel_requested = (1 << 63) | N
+    /// - Signal handler checks: (cancel_requested & 0x7FFFFFFFFFFFFFFF) == current_generation
+    cancel_requested: AtomicU64,
+
+    /// Flag indicating whether a guest function call is currently in progress.
+    ///
+    /// **true**: A guest function call is active (between call start and completion)
+    /// **false**: No guest function call is active
+    ///
+    /// # Purpose
+    ///
+    /// This flag prevents kill() from having any effect when called outside of a
+    /// guest function call. This solves the "kill-in-advance" problem where kill()
+    /// could be called before a guest function starts and would incorrectly cancel it.
+    ///
+    /// # Behavior
+    ///
+    /// - Set to true at the start of call_guest_function_by_name_no_reset()
+    /// - Cleared at the end of call_guest_function_by_name_no_reset()
+    /// - kill() only stamps cancel_requested if call_active is true
+    /// - If kill() is called when call_active=false, it returns false and has no effect
```

Comment:
Is this actually the case? I don't see it returning early when call is not active.